core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -173,51 +255,67 @@ private[core] object SchedulerLogic {
     *   - Prune a terminal / unknown task status.
     * - If a podSpec is marked as terminal, then issue a kill.
     */
-  private[core] def computeNextStateForPods(specs: SpecificationState, state: SchedulerLogicState)(changedPodIds: Set[PodId]): SchedulerLogicIntents = {
-    changedPodIds.foldLeft(SchedulerLogicIntentsBuilder.empty) { (initialIntents, podId) =>
-      specs.podSpecs.get(podId) match {
-        case None =>
-          // TODO - this should be spurious if the podStatus is non-terminal
-          def maybePrunePodStatus(effects: SchedulerLogicIntentsBuilder) = {
-            val existingTerminalStatus = state.podStatuses.get(podId).exists(terminalOrUnreachable(_))
-            if (existingTerminalStatus) {
-              effects.withPodStatus(podId, None)
-            } else {
-              effects
+  private[core] def computeNextStateForPods(specs: SpecificationState, state: SchedulerLogicState)(
+      changedPodIds: Set[PodId]): SchedulerLogicIntents = {
+    import ProtoConversions._
+    changedPodIds
+      .foldLeft(SchedulerLogicIntentsBuilder.empty) { (initialIntents, podId) =>
+        specs.podSpecs.get(podId) match {
+          case None =>
+            // TODO - this should be spurious if the podStatus is non-terminal
+            def maybePrunePodStatus(effects: SchedulerLogicIntentsBuilder) = {
+              val existingTerminalStatus = state.podStatuses.get(podId).exists(terminalOrUnreachable(_))
+              if (existingTerminalStatus) {
+                effects.withPodStatus(podId, None)
+              } else {
+                effects
+              }
             }
-          }
 
-          def maybePruneRecord(effects: SchedulerLogicIntentsBuilder) = {
-            if (state.podRecords.contains(podId)) {
-              // delete podRecord
-              effects.withPodRecord(podId, None)
-            } else {
-              effects
+            def maybePruneRecord(effects: SchedulerLogicIntentsBuilder) = {
+              if (state.podRecords.contains(podId)) {
+                // delete podRecord
+                effects.withPodRecord(podId, None)
+              } else {
+                effects
+              }
             }
-          }
 
-          maybePrunePodStatus(maybePruneRecord(initialIntents))
+            maybePrunePodStatus(maybePruneRecord(initialIntents))
 
-        case Some(podSpec) =>
-          if (podSpec.goal == Goal.Terminal) {
-            taskIdsFor(podSpec).foldLeft(initialIntents) { (effects, taskId) =>
-              effects.withMesosCall(Mesos.Call.Kill(taskId))
+          case Some(podSpec) =>
+            if (podSpec.goal == Goal.Terminal) {
+              taskIdsFor(podSpec).foldLeft(initialIntents) { (effects, taskId) =>
+                effects.withMesosCall(
+                  mesosCallFactory.newKill(taskId.asProto, state.podRecords.get(podSpec.id).map(_.agentId.asProto), None))
```

Comment:
is it cool as we're initiating kill for any change? E.g. the change is emitted even on a snapshot. Or we're going to tackle this later?

Reply:
Yeah, i think it is fine for now. We'll want a more robust kill mechanism with retry I think in the future perhaps? We should add a JIRA

Reply:
https://jira.mesosphere.com/browse/DCOS-48513